sql: retry more distributed errors as local

yuzefovich opened the pull request:
This PR contains a couple of commits that increase the allow-list of errors that are retried locally. In particular, it allows us to hide some issues we have around using DistSQL and shutting down SQL pods.

Fixes: #106537.
Fixes: #108152.
Fixes: #108271.

Release note: None

yuzefovich committed:
sql: retry all DistSQL runner dial errors

This commit marks the error that DistSQL runners produce when dialing
remote nodes in a special way that is now always retried-as-local. In
particular, this allows us to fix two problematic scenarios that could
occur when using secondary tenants:
- when attempting to start a pod with stale instance information
- the port is in use by an RPC server for the same tenant, but with
a new instance id.

This commit includes the test from Jeff that exposed the gap in the
retry-as-local mechanism.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108336)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/sql/distsql_running.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbH6kBo0zrtwF0yaTKa:-NbH6kBo0zrtwF0yaTKb:b-60aso5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c2c676c74ad2d61b94ace8327fc263bf08b6e6a/pkg/sql/distsql_running.go#L107)):*
> ```Go
> }
> 
> func isDialErr(err error) bool {
> ```

~I briefly looked into using `errors.Is` or `errors.As` here but wasn't sure what's the canonical way that would definitely work. Probably Raphael has a suggestion here.~ Reimplemented with `errors.HasType`.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
server: minor cleanup of logging around draining

This commit moves the "drain succeeded" logging message to be at the
very end of the drain process. Also, it removes now stale comment.

Release note: None

yuzefovich committed:
sql: retry all "rpc error" distributed errors as local

This commit includes all errors that contain `rpc error` substring to be
retried-as-local. In particular, this allows us to avoid problems with
DistSQL using no-longer-live SQL pod after that pod is shutdown. (This
usage of the downed pod is currently expected given that the cache of
live instances isn't updated when the pod is shutdown.)

Release note: None

yuzefovich reviewed (commented):
I added another commit to include more errors in the allow-list for retrying. Done pushing for now, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbI7-Z8D_j__vL5L-uJ:b8zh4l0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbKFJiRANWT0x1acqZH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @yuzefovich)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKEthb0p5pVF5sanb7:b2kjo9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
> ```Go
> 	}()
> 
> 	listener, err := net.Listen("tcp", rpcAddr)
> ```

Why is this listener required?


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

JeffSwenson commented on pkg/sql/distsql_running.go:
```diff
@@ -100,6 +101,22 @@ type runnerResult struct {
 	err    error
 }
 
+type runnerDialErr struct {
```
I like the approach of tagging errors that occur during the dial flow. This is a robust way to handle issues that occur during start up, which is the most common case we have seen for Serverless.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson and @knz)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKqqpGAx-4DglyikTr:bcoy4qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why is this listener required?
</blockquote></details>

I'm not sure, I just copied this test from Jeff #106538. @JeffSwenson can you share some context about this listener?


<!-- Sent from Reviewable.io -->

JeffSwenson commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson and @knz)
> 
> _[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKqqpGAx-4DglyikTr:bcoy4qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):_
> 
> Previously, mgartner (Marcus Gartner) wrote…
> I'm not sure, I just copied this test from Jeff #106538. @JeffSwenson can you share some context about this listener?

The test is covering the following scenario:
1. A sql server starts up and is assigned port `a`
2. The sql server shuts down and releases port `a`
3. Something else starts up and claims port `a`. In the test that is the listener. This is important because the listener causes connections to `a` to hang instead of responding with a RESET packet.
4. A different server with stale instance information schedules a distsql flow and attempts to dial port `a`.

yuzefovich reviewed (commented):
Usage of the string matching for errors in `IsDistSQLRetryable` is not great, but it's not a new problem and we do have #82847 as a tracking issue to improve the situation.

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbMLnXiE1W_J2vFKvwb:b-nvw0p8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbMLhvM0T3pl5mw1aLT:b-hm7rxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

The test is covering the following scenario:
1. A sql server starts up and is assigned port `a`
2. The sql server shuts down and releases port `a`
3. Something else starts up and claims port `a`. In the test that is the listener. This is important because the listener causes connections to `a` to hang instead of responding with a RESET packet.
4. A different server with stale instance information schedules a distsql flow and attempts to dial port `a`.

</blockquote></details>

Thanks for this context. I incorporated it as a comment on the test.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Hm, I'm confused
```
[23:12:22][Run unit tests] ERROR: /go/src/github.com/cockroachdb/cockroach/pkg/jobs/BUILD.bazel:88:8: GoCompilePkg pkg/jobs/autoconfig.recompile1376.a failed: (Exit 1): builder failed: error executing command (from target //pkg/jobs:jobs_test) bazel-out/k8-opt-exec-2B5CBBC6/bin/external/go_sdk/builder_reset/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -tags bazel,gss,bazel,gss -src pkg/server/autoconfig/auto_config.go ... (remaining 62 arguments skipped)
[23:12:22][Run unit tests] 
[23:12:22][Run unit tests] Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
[23:12:22][Run unit tests] compilepkg: panic: runtime error: invalid memory address or nil pointer dereference
[23:12:22][Run unit tests] [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5e89bc]
```
Let's try one more time.

bors r+

yuzefovich commented:
Known failure #108340.

bors r+

yuzefovich commented:
The latest failure is at least legitimate:
```
=== RUN   TestStartTenantWithStaleInstance
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
    test_log_scope.go:81: use -show-logs to present logs inline
    server_sql_test.go:450: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
        	Error:      	Received unexpected error:
        	            	listen tcp 127.0.0.1:45845: bind: address already in use
        	Test:       	TestStartTenantWithStaleInstance
```
I'll take a look tomorrow, but I did stress this new test on the gceworker with no failures.

mgartner commented:
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbQZGDICh6DoJx3ilay:bmgfde0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks for this context. I incorporated it as a comment on the test.
</blockquote></details>

Thanks for the context. The only part I'm missing is how the "different server" is dialing `rpcAddr` - I don't see how that's forced.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbSD8Bq7gOsLlguXVtM:bzh9mgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):*
> ```Go
> 
> 	listener, err := net.Listen("tcp", rpcAddr)
> 	require.NoError(t, err)
> ```

For the failure that I got under stress on CI on this line:
```
=== RUN   TestStartTenantWithStaleInstance
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
    test_log_scope.go:81: use -show-logs to present logs inline
    server_sql_test.go:450: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
        	Error:      	Received unexpected error:
        	            	listen tcp 127.0.0.1:45845: bind: address already in use
        	Test:       	TestStartTenantWithStaleInstance
```
it seems like the stopped tenant hasn't released the socket, and to me it seems like a benign error with the test setup. I'm inclined to introduce SucceedsSoon until `net.Listen` doesn't return an error. @JeffSwenson WDYT?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbSGeRO3hhdz09uXQ05:bc5bdmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Thanks for the context. The only part I'm missing is how the "different server" is dialing `rpcAddr` - I don't see how that's forced.
</blockquote></details>

I think I see this. In step 4. the "different server" is the new being-started-right-now SQL pod (which itself will listen on a different port), and when starting up, it sees that there is another SQL pod for the same tenant in `sql_instances` system table (see https://github.com/cockroachdb/cockroach/issues/106537#issue-1797296642). Thus, it attempts to dial that SQL pod to perform the startup migration, which fails because we have `listener` not responding to that dial attempt.


<!-- Sent from Reviewable.io -->

JeffSwenson commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)
> 
> _[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbSD8Bq7gOsLlguXVtM:bzh9mgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):_
> 
> > ```go
> > 	listener, err := net.Listen("tcp", rpcAddr)
> > 	require.NoError(t, err)
> > ```
> 
> For the failure that I got under stress on CI on this line:
> 
> ```
> === RUN   TestStartTenantWithStaleInstance
>     test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
>     test_log_scope.go:81: use -show-logs to present logs inline
>     server_sql_test.go:450: 
>         	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
>         	Error:      	Received unexpected error:
>         	            	listen tcp 127.0.0.1:45845: bind: address already in use
>         	Test:       	TestStartTenantWithStaleInstance
> ```
> 
> it seems like the stopped tenant hasn't released the socket, and to me it seems like a benign error with the test setup. I'm inclined to introduce SucceedsSoon until `net.Listen` doesn't return an error. @JeffSwenson WDYT?

Adding a `SucceedsSoon` LGTM. Another option is we could start the listener, then inject a synthetic sql_instance row. But I like the current implementation of the test because it is very generic and has minimal coupling to the values in the sql_instance table.

yuzefovich reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbV7DIPDY4bsmKM3Ja1:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbV7AQxD5S5uBZ8gF5b:bcxgpet) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Adding a `SucceedsSoon` LGTM. Another option is we could start the listener, then inject a synthetic sql_instance row. But I like the current implementation of the test because it is very generic and has minimal coupling to the values in the sql_instance table.
</blockquote></details>

Thanks, let's keep `SucceedsSoon` then.


<!-- Sent from Reviewable.io -->
